bls/bls.go

```diff
@@ -1,79 +1,110 @@
+	PublicKeyLength = bls12381.SizeOfG1AffineCompressed
+	SecretKeyLength = fr.Bytes
+	SignatureLength = bls12381.SizeOfG2AffineCompressed
 )
 
 type (
-	PublicKey = bls12381.PointG1
-	SecretKey = bls12381.Fr
-	Signature = bls12381.PointG2
+	PublicKey = bls12381.G1Affine
+	SecretKey = fr.Element
+	Signature = bls12381.G2Affine
 )
 
 var (
+	_, _, g1One, _            = bls12381.Generators()
+	domain                    = []byte("BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_POP_")
 	ErrInvalidPubkeyLength    = errors.New("invalid public key length")
 	ErrInvalidSecretKeyLength = errors.New("invalid secret key length")
 	ErrInvalidSignatureLength = errors.New("invalid signature length")
 	ErrSecretKeyIsZero        = errors.New("invalid secret key is zero")
+	ErrPointNotOnCurve        = errors.New("point not on the curve")
+	ErrPointNotInSubGroup     = errors.New("point not in correct subgroup")
 )
 
 func PublicKeyToBytes(pk *PublicKey) []byte {
-	return bls12381.NewG1().ToCompressed(pk)
+	pkBytes := pk.Bytes()
+	return pkBytes[:]
 }
 
 func SecretKeyToBytes(sk *SecretKey) []byte {
-	return sk.ToBytes()
+	skBytes := sk.Bytes()
+	return skBytes[:]
 }
 
 func SignatureToBytes(sig *Signature) []byte {
-	return bls12381.NewG2().ToCompressed(sig)
+	sigBytes := sig.Bytes()
+	return sigBytes[:]
 }
 
 func PublicKeyFromBytes(pkBytes []byte) (*PublicKey, error) {
 	if len(pkBytes) != PublicKeyLength {
 		return nil, ErrInvalidPubkeyLength
 	}
-	return bls12381.NewG1().FromCompressed(pkBytes)
-}
-
-func PublicKeyFromSecretKey(sk *SecretKey) (*PublicKey, error) {
-	if sk.IsZero() {
-		return nil, ErrSecretKeyIsZero
-	}
 	pk := new(PublicKey)
-	g1 := bls12381.NewG1()
-	g1.MulScalar(pk, &bls12381.G1One, sk)
-	return pk, nil
+	err := pk.Unmarshal(pkBytes)
+	if !pk.IsOnCurve() {
+		return nil, ErrPointNotOnCurve
+	}
+	if !pk.IsInSubGroup() {
+		return nil, ErrPointNotInSubGroup
+	}
+	return pk, err
 }
 
 func SecretKeyFromBytes(skBytes []byte) (*SecretKey, error) {
 	if len(skBytes) != SecretKeyLength {
 		return nil, ErrInvalidSecretKeyLength
 	}
-	sk := bls12381.NewFr().FromBytes(skBytes)
+	sk := new(SecretKey).SetBytes(skBytes)
 	if sk.IsZero() {
 		return nil, ErrSecretKeyIsZero
 	}
 	return sk, nil
 }
 
+func SignatureFromBytes(sigBytes []byte) (*Signature, error) {
+	if len(sigBytes) != SignatureLength {
+		return nil, ErrInvalidSignatureLength
+	}
+	sig := new(Signature)
+	err := sig.Unmarshal(sigBytes)
+	if !sig.IsOnCurve() {
+		return nil, ErrPointNotOnCurve
+	}
+	if !sig.IsInSubGroup() {
+		return nil, ErrPointNotInSubGroup
+	}
+	return sig, err
+}
+
 func GenerateRandomSecretKey() (*SecretKey, error) {
-	return new(SecretKey).Rand(rand.Reader)
+	return new(SecretKey).SetRandom()
+}
+
+func PublicKeyFromSecretKey(sk *SecretKey) (*PublicKey, error) {
+	if sk.IsZero() {
+		return nil, ErrSecretKeyIsZero
+	}
+	skBigInt := new(big.Int)
+	sk.BigInt(skBigInt)
+	pk := new(bls12381.G1Affine).ScalarMultiplication(&g1One, skBigInt)
+	if !pk.IsOnCurve() {
```

Comment:
We can avoid the IsOnCurve and Subgroup checks, since we did a ScalarMul with the generator, we know the result will be in the subgroup and on the curve